### DISCUSSION TITLE
start arduino script when object is visible in camera

### DISCUSSION THREAD

**Discussion by salvatoreNRIM** [bonsai-rx/bonsai#1142] [Q&A]
Hi Bonsai community!

I'd like to use Bonsai to trigger an LED (for optogenetic activation), for 1 second after the mouse is visible in the camera. This trigger should only happen once per trial (so I used a delay of 5 sec, for preventing longer stimulations). To do so I start a script in Arduino (attached) which makes the pulse of desired width, and then I start the bonsai program (also attached). However the Bonsai script does not seem to affect the light pulse. Would anyone have some suggestions on how to implement this?
[Trigger_optogenetic_ridge_task.zip](https://github.com/bonsai-rx/bonsai/files/9999095/Trigger_optogenetic_ridge_task.zip)
[LED_optogenetic_blink.zip](https://github.com/bonsai-rx/bonsai/files/9999116/LED_optogenetic_blink.zip)

**Comment by bruno-f-cruz** [comment]
Hey, 
Have you uploaded Firmata to the Arduino?
If so, can you try the attached workflow? 
cheers,
B
[Trigger_optogenetic_ridge_task.zip](https://github.com/bonsai-rx/bonsai/files/10001015/Trigger_optogenetic_ridge_task.zip)

**Comment by salvatoreNRIM** [comment]
Hi @bruno-f-cruz, thanks for the quick reply. If I upload the firmata, and then run the bonsai script you sent me I only get a continuous pulse, while I want to get a train of pulses, as I wrote in the arduino code I attached in my first post:

```
`// the loop function runs over and over again forever
void loop() {
  digitalWrite(LED, HIGH);   // turn the LED on (HIGH is the voltage level)
  delay(15);                       // wait for 15 ms
  digitalWrite(LED, LOW);    // turn the LED off by making the voltage LOW
  delay(35);                       // wait for 35 ms
}`
```

Then I want to use Bonsai to trigger this arduino script for one second, stop it, and restart the trial.

**Reply by bruno-f-cruz** [reply]
Assuming you have firmata loaded, the workflow I sent you will, upon detecting a threshold crossing, turn the pin On for 1 second, Turn it off, and wait 5 secs to re-arm the whole system and listen for another threshold crossing.  Is this not the behavior you are looking for?

**Reply by salvatoreNRIM** [reply]
@bruno-f-cruz I need to load the arduino script I wrote (LED_optogenetic_blink) to generate a train of 15 ms pulses. After that, I loaded the firmata (which stops the train of pulses), and then I started the bonsai script you sent. This will now generate a continuous pulse for 1 sec, turn it off etc.. However, what I want to do is to generate a train of 15 ms pulses at 20 Hz for 1 sec, turn it off for 5 sec etc... I wrote the script in Arduino because I am not sure Bonsai can generate 15 ms pulses accurately, so I wanted to use Bonsai to start the script LED_optogenetic_blink for 1 sec upon detection of an object under the camera.

**Reply by bruno-f-cruz** [reply]
Let me clear something up. Once you load Firmata your previous script is overwritten. Additionally, you are correct, I would not trust Bonsai to generate those small delays faithfully. 

The solution might be to do something like coding your own interface in Arduino:
https://groups.google.com/g/bonsai-users/c/0-eIYPe8rKo/m/11chWHvOBgAJ

**Reply by salvatoreNRIM** [reply]
Thanks, yes that makes sense! I tried to adapt the script from the post you sent to my Bonsai script, but I run into the following issue. Would you have any suggestions on what the issue is? 
```
"Bonsai.WorkflowRuntimeException: The given port name does not start with COM/com or does not resolve to a valid serial port.
Parameter name: portName ---> System.ArgumentException: The given port name does not start with COM/com or does not resolve to a valid serial port.
Parameter name: portName
   at System.IO.Ports.SerialStream..ctor(String portName, Int32 baudRate, Parity parity, Int32 dataBits, StopBits stopBits, Int32 readTimeout, Int32 writeTimeout, Handshake handshake, Boolean dtrEnable, Boolean rtsEnable, Boolean discardNull, Byte parityReplace)
   at System.IO.Ports.SerialPort.Open()
   at Bonsai.IO.SerialPortManager.ReserveConnection(String portName, SerialPortConfiguration serialPortConfiguration)
   at Bonsai.IO.SerialStringWrite.<>c__DisplayClass8_0`1.<Process>b__0()
   at System.Reactive.Linq.ObservableImpl.Using`2._.Run()
   --- End of inner exception stack trace ---"
```
[trigger_opto_press_S.zip](https://github.com/bonsai-rx/bonsai/files/10036712/trigger_opto_press_S.zip)

**Reply by bruno-f-cruz** [reply]
You should not use the same COM port as a generic Serial device and as an "Arduino" device. Delete all references to one of them.

**Reply by salvatoreNRIM** [reply]
I deleted the COM part as serial device, but I think I still run into the same issue. I tried to use as a reference the example you wrote on this post
[https://groups.google.com/g/bonsai-users/c/bh6oosVIFXM](url)
[trigger_opto_press_S (2).zip](https://github.com/bonsai-rx/bonsai/files/10051382/trigger_opto_press_S.2.zip)

**Reply by bruno-f-cruz** [reply]
can you add the script you are uploading to Arduino? thanks

**Reply by salvatoreNRIM** [reply]
Sure, here you are!

[LED_optogenetic_with_s.zip](https://github.com/bonsai-rx/bonsai/files/10054858/LED_optogenetic_with_s.zip)

**Reply by bruno-f-cruz** [reply]
In the `.ino` you sent, you are waiting for the character "s" (ASCII code = 115) to be sent to the Arduino and then you will start blinking the LED forever. I don't think this is what you want based on your description. I have re-written the Arduino and the Bonsai files, let me know if it makes sense. I don't have a way to test it, but I think it will get you closer to what you want. You can change the logic in the arduino side to accommodate more complicated logic. For instance you could have one character to start ("s") and another one to finish the stimulation ("e").

Cheers,
Bruno

[opto_serial_arduino.zip](https://github.com/bonsai-rx/bonsai/files/10055846/opto_serial_arduino.zip)

**Reply by salvatoreNRIM** [reply]
Okay so I tried it with the files you sent me, and I get an error from Arduino 
`C:\Users\MOTION~1\AppData\Local\Temp\arduino_modified_sketch_350025\OptoArduinoSerial.ino.ino:2:26: warning: invalid conversion from 'const char*' to 'char' [-fpermissive]
 char trigger_character = "s";
`
I tried to change this line in Arduino to single apostrophe, but it didn't help 
`char trigger_character = "s";
`
Not sure whether it matters, but I am using Arduino 1.8.19

**Reply by bruno-f-cruz** [reply]
oops, try single quotes `'s'` instead of double quotes `"s"`.

**Reply by salvatoreNRIM** [reply]
The single quote get rid of the error in arduino, but unfortunately it still does not work: I first upload the arduino file, then run the Bonsai script, and although the object is detected properly the LED does not do anything. I tested the connection between arduino and the LED and it's working fine. I also tried to turn on the LED from the arduino directly and it's working as well. So I suspect the issue might be in the Bonsai script.

**Reply by bruno-f-cruz** [reply]
Can you turn on the LED by sending a "s" via the arduino serial console directly?

**Reply by salvatoreNRIM** [reply]
Oh yeah that's a good point. The LED does not turn on by pressing s on the keyboard, I tried to replace s with the ASCII character value as well in the arduino script, but it does not work either...

**Reply by bruno-f-cruz** [reply]
I will try to get my hands on an Arduino and update you then.

**Reply by salvatoreNRIM** [reply]
Thank you very much, I really appreciate it!

**Reply by bruno-f-cruz** [reply]
I wrongly swapped the increment and condition in the loop.

try this:

```C
void optogenetics_stimulation(int n, int on, int off){
  for (int i = 0 ; i<n; i++){
    digitalWrite(LED, 1);
    delay(on);
    digitalWrite(LED, 0);
    delay(off);
  }
}
```

Tested on my end and it seems to work

**Reply by salvatoreNRIM** [reply]
Yes it works perfectly. Thanks again!

**Reply by msidikpr** [reply]
Hello I am trying to do something similar but get an access to com port is denied error. It seems they got around it in this thread (https://groups.google.com/g/bonsai-users/c/0-eIYPe8rKo/m/11chWHvOBgAJ) but the solution was not posted.

**Reply by bruno-f-cruz** [reply]
can you share the smallest workflow that reproduces the issue? thanks
